Userland/Libraries/LibGL/Buffer.cpp

```diff
@@ -11,32 +12,92 @@ namespace GL {
 
 void GLContext::gl_bind_buffer(GLenum target, GLuint buffer)
 {
-    // FIXME: implement me
-    dbgln_if(GL_DEBUG, "{}({:#x}, {})): unimplemented", __FUNCTION__, target, buffer);
+    RETURN_WITH_ERROR_IF(target != GL_ARRAY_BUFFER && target != GL_ELEMENT_ARRAY_BUFFER, GL_INVALID_ENUM);
+
+    Buffer** target_buffer = target == GL_ELEMENT_ARRAY_BUFFER ? &m_element_array_buffer : &m_array_buffer;
+    RefPtr<Buffer> found_buffer;
+
+    if (buffer == 0) {
+        // If buffer is set to zero, we should unbind the currently bound buffer
+        *target_buffer = nullptr;
+    } else {
+        auto it = m_allocated_buffers.find(buffer);
+        if (it != m_allocated_buffers.end()) {
```

Comment:
This should be implemented:

>  GL_INVALID_VALUE is generated if buffer is not a name previously returned from a call to glGenBuffers.

Reply:
I noticed that this is not implemented in `Texture.cpp` either, maybe `NameAllocator` could be refactored to contain common code for creating/deleting OpenGL objects?

Reply:
I believe that check was left out from texture allocation code because old clients depends on the implicit "create when it does not exist" behavior. I'm not sure if that's the case for buffer creation as well?

Reply:
The documentation for `glBindBuffer` says "If no buffer object with name buffer exists, one is created with that name," although it does say later that `glGenBuffers` must be used to generate unused buffer names. I'm not sure if anything relies on this behaviour for buffers, but it seems like the best solution is to generate an error like the documentation says.

Reply:
Maybe we're doing this wrong for textures as well - not sure. There is a difference between _buffer name_ (or texture name) and _buffer object_ (or texture object). The _name_ must have been generated by a call to `glGen*`, but the _object_ is created if it isn't mapped by the (existing) name.